src/server/lock_manager/waiter_manager.rs

```diff
@@ -242,93 +433,80 @@ impl WaiterManager {
+        let f = waiter.on_timeout(move || {
+            wait_table
+                .borrow_mut()
+                .remove_waiter(lock, waiter_ts)
+                .and_then(|waiter| {
+                    detector_scheduler.clean_up_wait_for(waiter.start_ts, waiter.lock);
+                    waiter.notify();
+                    Some(())
                 });
-            handle.spawn(timer);
-        }
+        });
+        self.wait_table
+            .borrow_mut()
+            .add_waiter(waiter)
+            .and_then(|old| {
+                old.notify();
+                Some(())
+            });
+        handle.spawn(f);
     }
 
-    fn handle_wake_up(
-        &mut self,
-        handle: &Handle,
-        lock_ts: TimeStamp,
-        hashes: Vec<u64>,
-        commit_ts: TimeStamp,
-    ) {
-        let mut ready_waiters = self
-            .wait_table
-            .borrow_mut()
-            .take_ready_waiters(lock_ts, hashes);
-        ready_waiters.sort_unstable_by_key(|waiter| waiter.start_ts);
-
-        for (i, waiter) in ready_waiters.into_iter().enumerate() {
-            self.detector_scheduler
-                .clean_up_wait_for(waiter.start_ts, waiter.lock);
-            if self.wake_up_delay_duration > 0 {
-                // Sleep a little so the transaction with small start_ts will more likely get the lock.
-                let when = Instant::now()
-                    + Duration::from_millis(self.wake_up_delay_duration * (i as u64));
-                let timer = Delay::new(when)
-                    .and_then(move |_| {
-                        wake_up_waiter(waiter, commit_ts);
-                        Ok(())
-                    })
-                    .map_err(|e| info!("wake-up timer delay errored"; "err" => ?e));
-                handle.spawn(timer);
-            } else {
-                wake_up_waiter(waiter, commit_ts);
-            }
+    fn handle_wake_up(&mut self, lock_ts: TimeStamp, hashes: Vec<u64>, commit_ts: TimeStamp) {
+        let mut wait_table = self.wait_table.borrow_mut();
+        let new_timeout = Instant::now() + Duration::from_millis(self.wake_up_delay_duration);
+        for hash in hashes {
+            let lock = Lock { ts: lock_ts, hash };
+            // Notify the oldest one immediately.
+            wait_table
+                .remove_oldest_waiter(lock)
+                .and_then(|mut waiter| {
+                    self.detector_scheduler
+                        .clean_up_wait_for(waiter.start_ts, waiter.lock);
+                    waiter.conflict_with(lock_ts, commit_ts);
+                    waiter.notify();
+                    Some(())
+                });
+            // Others will be waked up after `wake_up_delay_duration`.
+            //
+            // NOTE: Actually these waiters are waiting for an unknown transaction.
+            // If there is a deadlock between them, it will be detected after timeout.
+            wait_table.get_mut(&lock).and_then(|waiters| {
```

Comment:
Can we avoid this look up by return waiters from the first look up?

Reply:
Of course.

Reply:
I can't make it. I'm defeated by the borrow checker. ..